mtt/src/mtt_types.rs

```diff
@@ -94,27 +93,512 @@ impl MttL2EntryType {
+    InvalidL1Entry,
+    /// The MTT L2 entries for the specified 1GB range failed to pass the invariant that all
+    /// 16 consecutive entries (each mapping 64MB) must have the same type.
+    ///  This should be treated as an operational error.
+    Invalid1GBRange,
+    /// The caller specified an operation on a 1GB range, but the L2 entry for the physical
+    /// address does not map a 1GB range.
+    Non1GBL2Entry,
+    /// The caller specified a operation on a 64MB range, but the L2 entry for the physical
+    /// address doesn't map a 64MB range.
+    Non64MBL2Entry,
+    /// The caller specified a operation on a 4K page, but the L2 entry for the physical
+    /// address doesn't map a 4K page.
+    Non4KL2Entry,
+    /// The caller specified an invalid physical address (too large for example)
+    InvalidAddress,
+    /// The specificied page size is yet supported
+    UnsupportedPageSize,
+    /// The caller specified an invalid number of pages for the range (example: 0, or would overflow address space)
+    InvalidAddressRange,
+    /// The platform initialization hasn't been completed. The implementation is expected
+    /// to initialize pointers to the MTT L2 and L1 page pool using platform-FW tables.
+    PlatformInitNotCompleted,
+    /// The platform configuration for the MTT L2 and L1 is invalid
+    InvalidPlatformConfiguration,
+    /// The platform didn't allocate sufficient L1 page pool pages to map the specified address
+    InsufficientL1Pages,
+}
+
+/// Holds the result of a MTT operation.
+pub type Result<T> = core::result::Result<T, Error>;
+
+const MTT_L1_ENTRY_COUNT: usize = 512;
+/// Struct to represent a L1 page
+/// Each 8-byte entry represents 32 4K pages, and the 512 total entries span a total of 64MB of address space.
+/// Each page uses a 2-bit encoding, with '00b as non-confidential and '01b as confidential ('11b is illegal).
+pub struct MttL1Page {
+    base: [u64; MTT_L1_ENTRY_COUNT],
+}
+struct PlatformMttTables {
+    mtt_l2_base: &'static mut [u64],
+    mtt_l1_page_pool_base: &'static mut [MttL1Page],
+}
+
+/// Holds references to the platform allocated L2 and L2 tables in a Mutex.
+/// This should be constructed by calling init() with the platform allocated backing
+/// memory for the L2 and L1 tables.
+pub struct PlatformMttInfo {
+    inner: Mutex<PlatformMttTables>,
+}
+
+// Singleton instance of PlatformMttInfo constructed by PlatformMttInfo::init()
+static PLATFORM_MTT_INFO: Once<PlatformMttInfo> = Once::new();
+
+// Utility structure for for MTT manipulation.
+// `current_l2_entry` and `current_phys_addr` are updated as we iterate
+// over the addresses in a range.
+struct MttHelper<'a> {
+    // Type of MTT L2 entry for the physical address in `current_phys_addr`
+    current_l2_entry: MttL2EntryType,
+    current_phys_addr: SupervisorPageAddr,
+    page_size: PageSize,
+    num_pages: usize,
+    inner: MutexGuard<'a, PlatformMttTables>,
+}
+
+impl PlatformMttInfo {
```

Comment:
I like the iterator for breaking up the range into the appropriate sizes, but I find the structure of the different types a bit odd. It seems more natural to me -- and looks a lot more like `PageTable` -- if the root structure (I guess that's `PlatformMttTables`?) holds the pointer to the MTT, and all the various "entry" types borrow that root structure. That root structure can also make the necessary guarantees for the `unsafe` blocks -- i.e. that all MTT pages (L1 & L2) are confidential, uniquely-owned, and not aliased -- since it validates the MTT at initialization and is solely responsible for modifying the MTT at runtime.

Reply:
I have tried to shuffle things around to try and make it match `PageTable`, but don't have an intuitive understanding of it like you do. So, we have `PlatformMttTables`, which maintains a Mutex on `PlatformMttInfo`, and is the guardian of the lock, so as to speak, and other entities borrow a reference to the internal structure by locking it as needed. We now use a `PageList<Page<InternalClean>>`  for the L1 pages, and the `unsafe` requirements have been moved to the `init()`.  In the other places where we deference the L1 page, we no longer have any specific requirements, since they were already satisfied upfront.